### PR TITLE
Ship agent skills for writing goldens

### PR DESCRIPTION
**Pull request by dmkasperski**
Adds three [Agent Skills](https://dart.dev/ai/package-skills) so an AI agent can write `golden_test` goldens properly, and documents them in the README.

## Distribution

They live in a top-level `skills/` directory, which is the layout dart.dev specifies for pub packages. That means consumers run:

```bash
dart run skills@ get
```

and the CLI finds them in their dependencies and installs into `.agents/skills/` — no copy-paste, and they update with the package. The directory names are prefixed with the package name (`golden_test-*`) because the tool silently skips skills that aren't.

Confirmed `skills/` is included in the published archive (723 KB → 764 KB, 0 warnings). Worth knowing: `.pubignore` excludes `.claude/`, so skills kept there would never have shipped.

| Skill | Scope |
|---|---|
| `golden_test-setup` | One-time setup: dependency, and a `flutter_test_config.dart` wired to the app's real themes, locales, fonts and DI |
| `golden_test-widget` | Component and design-system goldens |
| `golden_test-route` | Screen goldens |

Each keeps rules and decisions in `SKILL.md` and worked examples in `references/`.

## How the content was validated

Three end-to-end runs: an agent wrote goldens for a real app from scratch, working only from `lib/` and the skills, in an isolated copy with no git history to crib from. Each run's failures were fed back into the skills, then a fourth pass applied the revised skills to an existing suite.

The runs caught things that would otherwise have shipped:

- API described in the skills is checked against the package source, not recalled — one early draft documented a `disableAnimations` parameter that has never existed, and hand-rolled a helper the package already exports (`simulateRouteStack`).
- Guidance that produced waste: RTL goldens for an app with no RTL locale, a large-text golden on a phone-height canvas showing only a title, near-identical goldens from enumerating inputs rather than renderings.
- Internal contradictions between sections, and between a skill and its reference.
- Failure modes with no coverage: an overflow throws and produces *no* image, so "regenerate" is meaningless; a clipped golden passes and keeps passing; `--update-goldens` never deletes, so a renamed test orphans its PNG.

The skills also cover deriving edge cases from branches in the app's own formatting code, which is where the runs found real bugs rather than just missing coverage.

## Review round (a1b51fe)

Applied @LiLatee's four comments:

- `group('$FeatureScreen', …)` in the recipes — and a note that the same interpolation must **not** go in `goldenTest(name:)`, where the name is the PNG filename, so a class rename would silently rename every golden and orphan the old files.
- Each mock's construction, registration and stubbing kept together in the service-locator recipe rather than batched by operation, shown with two dependencies so the grouping is visible.
- Pull-to-refresh is now a `WidgetTester` extension, so the call site is one line.
- The `FLUTTER_TEST` flag is now bounded: it doesn't compile for web, isn't set when the app runs on a device under `integration_test`, and is read once per process. `FLUTTER_TEST=true` verified under `flutter test` on the VM. That thread is left open pending @LiLatee confirming whether the limitation he had in mind is one of these.

## Notes

- `SKILL.md` line counts are 500 / 492 / 537 against dart.dev's advisory 500-line guideline. Route is 37 over; recipes went to `references/` rather than cutting decision guidance to hit the number. Happy to trim further if you'd prefer it under.
- Every code snippet in the skills was compiled and run against the example app.
- Swept for project-specific content — no absolute paths, app names, or non-English text apart from a deliberate non-ASCII sample in the setup smoke test, which exists to prove `FontLoader` worked.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: skills/golden_test-route/references/patterns.md [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+unnecessary.
+
+### Service locator (GetIt or similar)
+
+For screens that resolve a dependency from a locator during construction or
+`initState`. **Scope registrations per test** — this is the leak the skill's
+§6 warns about:
+
+```dart
+group('FeatureScreen', () {
```
Comment (LiLatee) [inline skills/golden_test-route/references/patterns.md]: **suggestion:**

```suggestion
group('$FeatureScreen', () {
```

File: skills/golden_test-route/references/patterns.md [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+
+### Pull to refresh
+
+```dart
+action: (WidgetTester tester) async {
+  final center = tester.getCenter(find.byType(MaterialApp));
+  await tester.startGesture(center);
+  await tester.dragFrom(center, const Offset(0, 75));
+  await tester.pump();
+},
```
Comment (LiLatee) [inline skills/golden_test-route/references/patterns.md:433]: **suggestion:** Create a helper function to handle pull to refresh in a single line

File: skills/golden_test-route/references/patterns.md [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+```dart
+group('FeatureScreen', () {
+  late MockFeatureRepository repository;
+
+  setUp(() {
+    getIt.pushNewScope();
+    repository = MockFeatureRepository();
+    getIt.registerLazySingleton<FeatureRepository>(() => repository);
+    when(() => repository.dataStream).thenAnswer((_) => const Stream.empty());
+  });
```
Comment (LiLatee) [inline skills/golden_test-route/references/patterns.md:168]: **suggestion:** I am not sure if it's mentioned somewhere, but I am fan of grouping all related to a single mock actions togehter e.g.


```suggestion
  setUp(() {
    getIt.pushNewScope();
    
    repository = MockFeatureRepository();
    getIt.registerLazySingleton<FeatureRepository>(() => repository);
    when(() => repository.dataStream).thenAnswer((_) => const Stream.empty());
    
    repository2 = MockFeatureRepository2();
    getIt.registerLazySingleton<FeatureRepository2>(() => repository2);
    when(() => repository2.dataStream).thenAnswer((_) => const Stream.empty());
  });
```

instead of


```suggestion
  setUp(() {
    getIt.pushNewScope();
    
    repository = MockFeatureRepository();
    repository2 = MockFeatureRepository2();
    
    getIt.registerLazySingleton<FeatureRepository>(() => repository);
    getIt.registerLazySingleton<FeatureRepository2>(() => repository2);
    
    when(() => repository.dataStream).thenAnswer((_) => const Stream.empty());
    when(() => repository2.dataStream).thenAnswer((_) => const Stream.empty());
  });
```

File: skills/golden_test-setup/references/advanced-setup.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+Some non-determinism is awkward to fix through DI/mocking alone — a real
+map or ads SDK that doesn't work under `flutter test`, a debounce timer,
+a shimmer animation, "now". A small ambient flag, registered once and
+checked at those specific call sites, is often simpler than stubbing
+each one individually:
+
+```dart
+class TestEnvironment {
+  bool isRunningInTests = Platform.environment.containsKey('FLUTTER_TEST');
```
Comment (LiLatee) [inline skills/golden_test-setup/references/advanced-setup.md:129]: **info:** Didn't we talk about this at some point, that this solution doesn't always work? 🤔

**Review by LiLatee — APPROVED** [review]
+1 🐵

File: skills/golden_test-route/references/patterns.md (hunk 1 above)
Comment (dmkasperski) [reply skills/golden_test-route/references/patterns.md]: Applied in a1b51fe — the recipes now use `group('$FeatureScreen', …)`.

Worth a caveat that came out of this: the same interpolation is a trap in `goldenTest(name:)`, because that name *is* the PNG filename. A class rename would silently rename every golden and orphan the old files (`--update-goldens` only writes, never deletes). So: great in a group, avoid in `name:`. I added that to the route skill's naming section.

File: skills/golden_test-route/references/patterns.md (hunk 2 above)
Comment (dmkasperski) [reply skills/golden_test-route/references/patterns.md:433]: Good call — done in a1b51fe. It's now a `WidgetTester` extension:

```dart
extension PullToRefresh on WidgetTester {
  Future<void> pullToRefresh({double distance = 75}) async { … }
}
```

so the call site is `action: (tester) async => tester.pullToRefresh(),`. Compiled and ran it against the example app. Also noted to pump rather than settle, since `pumpAndSettle` times out on a refresh indicator that spins indefinitely.

File: skills/golden_test-route/references/patterns.md (hunk 3 above)
Comment (dmkasperski) [reply skills/golden_test-route/references/patterns.md:168]: Agreed, and it wasn't mentioned anywhere — now it is, in a1b51fe. The recipe groups per dependency (construct → register → stub), and I extended it to two mocks so the grouping is actually visible; with one it reads the same either way.

Your argument is the one I wrote down: it stays readable as dependencies multiply, and a mock missing a step is obvious. Batching by operation type means scanning three separate blocks to check one dependency is fully wired.

File: skills/golden_test-setup/references/advanced-setup.md (hunk 4 above)
Comment (dmkasperski) [reply skills/golden_test-setup/references/advanced-setup.md:129]: You're right, and thanks for pushing on it — I checked rather than trusting the section as written.

The base claim holds: `FLUTTER_TEST` is set to `"true"` by the `flutter test` runner, verified on the VM. But it's bounded in three ways the section didn't mention, and I've added them in a1b51fe:

- **It doesn't compile for web.** `dart:io` isn't available there, so an app also targeting web can't import `Platform` unconditionally — it needs a conditional import or a different signal.
- **It isn't set when the app runs on a real device**, so an `integration_test` suite driving the app on a device/emulator sees `false`. Usually what you want, but it means the flag is really "am I under `flutter test` on the VM", not "am I under test".
- **It's read once** — cached in a singleton it's a per-process constant.

So: fine for a widget/golden suite on the VM, poor as a general is-this-a-test switch. I also strengthened the closing advice — anything you *can* inject, inject, since a flag checked in production code is a branch that ships to users and a place the app can behave differently under test without anyone noticing.

Let me know if the limitation you were remembering was something else — happy to add it.
